src/adaptix/_internal/model_tools/introspection/sqlalchemy_tables.py

```diff
@@ -0,0 +1,149 @@
+from inspect import getfullargspec
+from typing import Any
+
+from sqlalchemy import inspect
+from sqlalchemy.sql.schema import CallableColumnDefault, ScalarElementColumnDefault
+
+from adaptix._internal.model_tools.definitions import (
+    DefaultFactory,
+    DefaultValue,
+    FullShape,
+    InputField,
+    InputShape,
+    NoDefault,
+    OutputField,
+    OutputShape,
+    Param,
+    ParamKind,
+    Shape,
+    create_attr_accessor,
+)
+from adaptix._internal.type_tools import get_all_type_hints
+
+
+class ColumnPropertyWrapper:
+    def __init__(self, column_property):
+        self.column_property = column_property
+
+
+def _is_context_sensitive(default):
+    try:
+        wrapped_callable = default.arg.__wrapped__
+    except AttributeError:
+        return True
+
+    spec = getfullargspec(wrapped_callable)
+    return len(spec.args) > 0
+
+
+def _get_sqlalchemy_type_for_column(column, type_hints):
+    try:
+        return type_hints[column.name].__args__[0]
+    except KeyError:
+        return column.type.python_type
+
+
+def _get_sqlalchemy_type_for_relationship(relationship, type_hints):
+    try:
+        return type_hints[str(relationship).split(".")[1]].__args__[0]
+    except KeyError:
+        return Any
+
+
+def _get_sqlalchemy_default(column_default):
+    if isinstance(column_default, CallableColumnDefault) and not _is_context_sensitive(column_default):
+        return DefaultFactory(factory=column_default.arg.__wrapped__)
+    if isinstance(column_default, ScalarElementColumnDefault):
+        return DefaultValue(value=column_default.arg)
+    return NoDefault()
+
+
+def _get_sqlalchemy_required(column):
+    return not (
+        column.default or column.nullable or column.server_default
+        or (column.primary_key and column.autoincrement and column.type.python_type is int)
```

Comment:
Why there is check of type?

Reply:
Because in any case `column.autoincrement` has `auto` value if `autoincrement` is not specified explicitly.